### PR TITLE
Observe surface data

### PR DESCRIPTION
**Pull request by geoffrey4444**
## Proposed changes

This PR 
  * factors existing checks that volume data is written correctly into a helper
  * adds an action (analogous to WriteReductionDataRow) suitable for writing volume data
  * adds a callback making use of this action to output surface "volume" data from the results of interpolation or an AH find
  * adds surface data output to the GH executables

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (knelli2) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: Is it possible to add the name of the 2D volume data file as a const_global_cache_tag like the name of the reduction file and 3D volume files and then just get it out of the cache? I'm wary of hard-coding the name here. Also, you don't document what the name of this file is anywhere.

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (knelli2) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: `pretty_type::name<InterpolationTargetTag>()`

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (knelli2) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: Is this namespace needed?

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (knelli2) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: `db::tag_name<Tag>()`

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (knelli2) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: I don't think `Parallel/Info.hpp` is used in this file.

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/IO/Observer/Tags.hpp
Comment (knelli2) [inline src/IO/Observer/Tags.hpp]: `SurfaceFileName`. Same below.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicWithHorizon.hpp
Comment (knelli2) [inline src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicWithHorizon.hpp]: You've already included this tag in the `ObserveSurfaceData` callback. Do you really need to include it here? Same for other GH executables.

File: src/IO/Observer/Tags.hpp
Comment (knelli2) [inline src/IO/Observer/Tags.hpp:248]: Add a test for this tag like with the volume and reduction tag.

**Review by knelli2 — CHANGES_REQUESTED** [review]
Looks good. You can squash in these changes with the rest directly.

File: src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicWithHorizon.hpp
Comment (geoffrey4444) [reply src/Evolution/Executables/GeneralizedHarmonic/EvolveGeneralizedHarmonicWithHorizon.hpp]: The executables failed to compile, complaining of missing tags, unless I included them in the executables as well

**Review by knelli2 — DISMISSED** [review]

File: tests/Unit/Helpers/IO/VolumeDataHelpers.hpp
Comment (nilsdeppe) [inline tests/Unit/Helpers/IO/VolumeDataHelpers.hpp]: To be consistent: `TestHelpers::io::volume_data` (or `VolumeData` at the end), then document `volume_data` namespace

File: tests/Unit/Helpers/IO/VolumeDataHelpers.hpp
Comment (nilsdeppe) [inline tests/Unit/Helpers/IO/VolumeDataHelpers.hpp]: promote to Doxygen comment

File: tests/Unit/Helpers/IO/VolumeDataHelpers.hpp
Comment (nilsdeppe) [inline tests/Unit/Helpers/IO/VolumeDataHelpers.hpp:28]: add doxygen comment

File: src/IO/Observer/VolumeActions.cpp
Comment (nilsdeppe) [inline src/IO/Observer/VolumeActions.cpp]: why version 4? At least a C++ comment would be helpful... Because normally I start at version 0 or 1...

File: src/IO/Observer/VolumeActions.hpp
Comment (nilsdeppe) [inline src/IO/Observer/VolumeActions.hpp]: ticks around `observers...`

File: src/IO/Observer/VolumeActions.hpp
Comment (nilsdeppe) [inline src/IO/Observer/VolumeActions.hpp]: ticks around `ObservationId`

File: src/IO/Observer/VolumeActions.hpp
Comment (nilsdeppe) [inline src/IO/Observer/VolumeActions.hpp]: ticks around `observers::ObserverWriter`

File: tests/Unit/IO/Observers/Test_VolumeObserver.cpp
Comment (nilsdeppe) [inline tests/Unit/IO/Observers/Test_VolumeObserver.cpp:339]: why not have this before the new parts you added? I'm confused on the split here

Can the new code be easily factored into a separate function so the test is easier to understand and that it's clear the test case covers 2 tests?

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (nilsdeppe) [inline tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: I don't love the idea of having one test that tests _all_ callbacks. Why is that the decision here? Can the test be factored into a test helper and then have 2 different test cases for the different callbacks use shared code from the helper? Generally we've kept a 1-to-1 correspondence between source files and test files. I think this makes the code easier to understand. E.g. if I want to add something to the one callback, I now need to understand a test for 2 callbacks

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: You're using these two types so you need to include them. Forward decls are only if you don't use a (pointer/reference) to a type

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp:65]: why not just inline the `theta_phi_points()` calls here? Then it's clear that `theta_phi` itself is never used

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [inline src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: It might be helpful to give some justification why, and/or how/if this can be extended. This should also be noted in the documentation

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (nilsdeppe) [inline tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: why not factor this into a separate function instead of a lambda? That would make the code easier to read, I think

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (nilsdeppe) [inline tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: why intermix the remove here with the added test?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/Helpers/IO/VolumeDataHelpers.hpp
Comment (geoffrey4444) [reply tests/Unit/Helpers/IO/VolumeDataHelpers.hpp]: Happy to make this change. By the way, I was just following what OberverHelpers.hpp does when I named this namespace. After looking across several test helpers domains, it looks to me like there isn't always consistency with the namespace names or with documenting them with doxygen

File: src/IO/Observer/VolumeActions.cpp
Comment (geoffrey4444) [reply src/IO/Observer/VolumeActions.cpp]: There's not a good reason, so I will switch this to 0. I thought our volume data version was 4, but I misunderstood that nonzero version numbers were only used for testing (Test_H5.cpp, e.g.).

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (geoffrey4444) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp:65]: I thought theta_phi_points() might be expensive, so I avoided calling it twice, but I'm happy to make this change

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (geoffrey4444) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: The generalization would be to output non-scalar components one at a time, treating each as a scalar. But in practice this isn't necessary, because we've never visualized vectors or higher-ranked tensors on a Strahlkorper, and I'm not sure how useful that would even be. I'll clarify the comment at line 89 along these lines.

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp:65]: Oh, I see. Could you make this not a `&` instead? That's what confused me. You're right, the result of `theta_phi_points()` is a value object not a reference, so there's some amount of computation done there

Comment (geoffrey4444) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp:65]: @nilsdeppe  just to confirm: you're saying no & line 73, right? Or did you mean no & on line 74 or somewhere else?

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (geoffrey4444) [reply tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: No reason...I'll just move this

File: tests/Unit/IO/Observers/Test_VolumeObserver.cpp
Comment (geoffrey4444) [reply tests/Unit/IO/Observers/Test_VolumeObserver.cpp:339]: Both files get written by the same process of running the mock action to interpolate (and then trigger the callbacks), so both must be checked for and removed before those actions run. It's simplest to check for these files existence at the start of the test, in my opinion. I added a comment explaining this

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (geoffrey4444) [reply tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: I took a stab at refactoring. I think it could be done, but it would be a nontrivial effort. It's not just precomputing some results...the mock framework stuff (components, mock actions, etc.) are mostly the same (other than the different callbacks), but the outputs and checks are different. I added a comment explaining this, and I rename the file to emphasize this is only checking two callbacks, not all of them. E.g., Kyle's callback to output AH centers is sufficiently different that it wouldn't be included here.

If you insist, I could spend the effort to refactor the test, but if you're OK with this (perhaps with revised wording in the comments I added), I'd rather spend the effort elsewhere

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp:65]: Yes, that's right, just line 73:
```cpp
     const std::array<DataVector, 2> theta_phi = ylm.theta_phi_points(); 
```

File: tests/Unit/Helpers/IO/VolumeDataHelpers.hpp
Comment (nilsdeppe) [reply tests/Unit/Helpers/IO/VolumeDataHelpers.hpp]: Yep, though we are trying to move to the "standard" :) another one of those things that, if I could time travel, I'd go back and start with the convention at the beginning :) Thanks for changing this!

File: src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp]: Sorry, I was thinking of expanding the `static_assert` message to include the reason why. That way people don't need to find the source code and hope there's a comment. I'm fine with having a longer C++ comment with a shorter `This could be generalized if needed, see the code comment above the assert.`

File: tests/Unit/IO/Observers/Test_VolumeObserver.cpp
Comment (nilsdeppe) [reply tests/Unit/IO/Observers/Test_VolumeObserver.cpp:339]: Sorry, these are the removals at the end of the test, not the beginning of the test. 

Ping on the question about factoring into a function that gets called. You can pass in the runner, etc. so no need to duplicate those. Just having a function would make the test easier to understand.

File: tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp
Comment (nilsdeppe) [reply tests/Unit/ParallelAlgorithms/Interpolation/Test_PostInterpolationCallbacks.cpp]: Okay, sounds good!

**Review by nilsdeppe — COMMENTED** [review]
Fixups looks good, a couple pings, please squash and rebase everything including new changes.

File: tests/Unit/IO/Observers/Test_VolumeObserver.cpp
Comment (geoffrey4444) [reply tests/Unit/IO/Observers/Test_VolumeObserver.cpp:339]: I think I got confused on the ping...I didn't realize this was a separate request. Yes, I can put the surface test itself into a function, to make that clear. I'll do that in the next fixup commit.

Update: OK, I've dobule-checked that the checks to remove the files are handled separately, because I moved all of the checks for the WriteVolumeData file into a function, as you requested here.

Comment (nilsdeppe) [reply tests/Unit/IO/Observers/Test_VolumeObserver.cpp:339]: Great! No need for a fixup, just squash right away so we don't have to wait on CI again :)

**Comment by geoffrey4444** [comment]
@nilsdeppe Thanks again for the review! I think I got everything this time, but please ping me if I missed anything

File: src/IO/Observer/VolumeActions.hpp
Comment (nilsdeppe) [inline src/IO/Observer/VolumeActions.hpp]: rename to `volume_file_lock`?

File: src/IO/Observer/Tags.hpp
Comment (nilsdeppe) [inline src/IO/Observer/Tags.hpp:240]: In the future, could you please keep commits like these smaller. E.g. add the tags in their own commit, and add the action in its own commit?

**Review by nilsdeppe — COMMENTED** [review]
LGTM, sorry I missed one naming thing previously. Please squash again...

File: src/IO/Observer/VolumeActions.hpp
Comment (geoffrey4444) [reply src/IO/Observer/VolumeActions.hpp]: 🤦‍♂️ thanks for catching this!

**Review by nilsdeppe — APPROVED** [review]
LGTM!

**Review by knelli2 — APPROVED** [review]
I'm happy with this. It will be a good diagnostic tool to have.

One thing to note though: Now, every time we do a horizon find using the events, we not only have one row of reduction data being written, but also some volume data (surface data). Hopefully, because it's only `Scalar`s, it won't be too much, but it's something we should keep in mind going forward. Doing an AH find every time step is now more costly in terms of disk space.

**Comment by geoffrey4444** [comment]
> I'm happy with this. It will be a good diagnostic tool to have.
> 
> One thing to note though: Now, every time we do a horizon find using the events, we not only have one row of reduction data being written, but also some volume data (surface data). Hopefully, because it's only `Scalar`s, it won't be too much, but it's something we should keep in mind going forward. Doing an AH find every time step is now more costly in terms of disk space.

Good point! I'd like to have a second set of horizon finders that runs less often than what's needed for the control system. We might want to dump surface data often near a time when the control system fails, but otherwise, probably having much less frequent output is fine, both for surface data and for physics reduction data, like masses and spins.

**Comment by geoffrey4444** [comment]
Two of my students are looking into testing this

**Comment by wthrowe** [comment]
I'm confused by the last few comments.  This PR doesn't appear to have anything to do with control system measurements.

**Comment by knelli2** [comment]
@geoffrey4444 The horizon finds using events ***are*** the ones that you would do less frequently. Just set the interval in the input file to like 1000 slabs or something. My statement was intended for those horizon finds triggered by events because somebody could, in theory, set the event trigger to run every time step which would dump a lot of surface data now. Control system measurements don't do any observing at all at the moment.

**Comment by nilsdeppe** [comment]
Before merging I'd like to make sure @wthrowe's confusion is clarified

**Comment by wthrowe** [comment]
That clarified it for me.  It makes sense now.  Thanks.

**Pull request merged**
